### PR TITLE
EKF2: Improvements and Solo Support

### PR DESCRIPTION
**Pull request by priseborough**
This replaces https://github.com/ArduPilot/ardupilot/pull/4239 and https://github.com/ArduPilot/ardupilot/pull/4237

**Testing**

Replay testing on the following logs:

http://uav.tridgell.net/ReplayLogs/
https://drive.google.com/open?id=0B0cCriOUIHS0eGNxTTYtcWxXN3c
https://drive.google.com/open?id=0B0cCriOUIHS0c3dHMWFvaUtxbDg

SITL testing using SiM_MAG_ALY_X,Y,Z  in the range +-300 to create up to  90 degrees of yaw error on ground.

SITL testing on plane with magnetometer disabled to verify that changes have not affected this capability. 

**Flight Testing**

Tests before recent rebase using 3DR Solo hardware:

https://drive.google.com/open?id=0B0cCriOUIHS0eGNxTTYtcWxXN3c
https://drive.google.com/open?id=0B0cCriOUIHS0c3dHMWFvaUtxbDg

Tests after rebase using  3DR Solo hardware. Iron bar placed adjacent to compass leg to induce a magnetic anomaly. 

https://drive.google.com/open?id=0B0cCriOUIHS0MnY0YzZ3QzlERlU

Initial and final field resets confirmed in data
screen shot 2016-06-20 at 1 50 36 pm

Seconds test with bar placed at a different orientation:

https://drive.google.com/open?id=0B0cCriOUIHS0WXE0SkJRdlBNTlU

No 'toilet bowling' observed and innovations were low after reset.

Landings next to the iron bar and taking off again showed expected switch to heading fusion when disarmed and re-initialisation of field states when airborne.

screen shot 2016-06-20 at 1 44 22 pm

### CODE REVIEW

**Comment by tridge** [comment]
hi Paul, I presume this needs more testing with the new SITL mag anomoly code?

**Comment by priseborough** [comment]
Yes, I have done some preliminary testing, but do some more with different anomaly levels this Monday.

**Comment by priseborough** [comment]
@tridge  I've completed functional testing. Some replay logs on different setups will help with fine tuning parameters, but that data will be easier to get once it is in master.

File: libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp [hunk 1]
```diff
@@ -20,42 +20,68 @@ extern const AP_HAL::HAL& hal;
...
-    bool hgtCheckPassed = (stateStruct.position.z  - posDownAtTakeoff) < -5.0f;
+    // and can perform a final reset of the yaw and field states
+    bool finalResetRequest = false;
+    if (flightResetAllowed && !assume_zero_sideslip()) {
+        finalResetRequest = (stateStruct.position.z  - posDownAtTakeoff) < -5.0f;
+    }
+
+    // Check for bad yaw casued by ground based magnetic anomaly
+    bool interimResetRequest = false;
+    if (flightResetAllowed && !assume_zero_sideslip()) {
```
Comment (OXINARF) [inline libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp]: Any reason for having the repeated if? Or was it not supposed to be the same if?

File: libraries/AP_NavEKF2/AP_NavEKF2_Outputs.cpp [hunk 2]
```diff
@@ -541,6 +548,9 @@ void NavEKF2_core::send_status_report(mavlink_channel_t chan)
     if (filt_state.flags.pred_horiz_pos_abs) {
         flags |= EKF_PRED_POS_HORIZ_ABS;
     }
+    if (filt_state.flags.gps_glitching) {
+        flags |= (1<<15);
```
Comment (OXINARF) [inline libraries/AP_NavEKF2/AP_NavEKF2_Outputs.cpp]: The flag should be added to MAVLink and then using the name here.

And maybe we should think about sending the generic ESTIMATOR_STATUS, but that's for another time.

File: libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp (hunk 1 above)
Comment (priseborough) [inline libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp]: they are separate checks and were originally using different top level entry conditions. I will merge them.

File: libraries/AP_NavEKF2/AP_NavEKF2_Outputs.cpp (hunk 2 above)
Comment (priseborough) [inline libraries/AP_NavEKF2/AP_NavEKF2_Outputs.cpp]: Not sure how to do that. I don't want to delay getting this in waiting to get a change to mavlink pushed through.

**Comment by priseborough** [comment]
Some more testing from this morning on a Solo;

https://drive.google.com/open?id=0B0cCriOUIHS0elU0MmQwWU1WMUE

**Comment by priseborough** [comment]
I rebased on latest master and pulled out the patch that adds the additional mavlink flag, so all review comments are now addressed.

**Comment by tridge** [comment]
applied, thanks Paul and Jon!

**Pull request closed without merging**